### PR TITLE
quant: add QAT fused Linear-Bn1d [1/x]: prepared module

### LINKED ISSUE #66483: Quantization - Linear-Bn1d fusion should support QAT
## 🐛 Bug

https://github.com/pytorch/pytorch/pull/50748 added Linear-Bn1d fusion for PTQ.  We should also have it for QAT.  This is useful for vision transformers.

high level steps to enable this:
1. add a nni.LinearBn1d module
2. add a nniqat.LinearBn1d module
3. create the mapping and the logic to swap from (1) to (2)
4. create the mapping and the logic from (2) to quantized.Linear module
5. test everything in eager and FX

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72796
* __->__ #72431

Summary:

Adds support for a fused QAT observed module for `Linear` followed by
`BatchNorm1d`. In this PR, only the support for prepared module with
fake_quants in the right places is added.

A future PR will add support for `convert`, and tests for eager and FX
graph mode workflows.

Similar to conv-bn, we rescale the weight before applying the fake
quant, and undo the rescaling after the linear operation.

Test Plan:

```
python test/test_quantization.py TestQuantizeEagerQATNumerics.test_linear_bn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34044427](https://our.internmc.facebook.com/intern/diff/D34044427)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34044427).

File: torch/nn/intrinsic/qat/modules/linear_fused.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
+import torch
+import torch.nn as nn
+import torch.nn.intrinsic as nni
+import torch.nn.functional as F
+from torch.nn import init
+from torch.nn.parameter import Parameter
+
+
+class LinearBn1d(nn.modules.linear.Linear, nni._FusedModule):
```
Comment (jerryzh168) [inline torch/nn/intrinsic/qat/modules/linear_fused.py:9]: nit: nn.modules.linear.Linear can be nn.Linear?

File: torch/nn/intrinsic/qat/modules/linear_fused.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                 in_features, out_features, bias=True,
+                 # BatchNorm1d args
+                 # num_features: out_features
+                 eps=1e-05, momentum=0.1,
+                 # affine: True
+                 # track_running_stats: True
+                 # Args for this module
+                 freeze_bn=False,
+                 qconfig=None):
+        nn.modules.linear.Linear.__init__(self, in_features, out_features, bias)
```
Comment (jerryzh168) [inline torch/nn/intrinsic/qat/modules/linear_fused.py:36]: same here

File: torch/nn/intrinsic/qat/modules/linear_fused.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        #
+        #   # scale the weight by previous batch's running statistics
+        #   scale_factor = bn.w / bn.running_std_from_prev_batch
+        #   # do the linear transformation without bias
+        #   x1_scaled = F.linear(x0, fq(w * scale_factor), 0)
+        #   # reverse the scaling and add original bias
+        #   x1_orig = x1_scaled / scale_factor + b
+        #   x2 = self.bn(x1_orig)
+
+        running_std = torch.sqrt(self.bn.running_var + self.bn.eps)
```
[the remark this answers is not in the document]
Comment (vkuzo) [reply torch/nn/intrinsic/qat/modules/linear_fused.py:101]: this is the same logic as fused QAT conv-bn

[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/nn/intrinsic/qat/modules/linear_fused.py:101]: oh OK, sounds good

**Review by jerryzh168 — APPROVED** [review]

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34044427).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34044427).
